Issue 7 instead of disabling local patches check entirely log issues as warnings and throw errors by default

beto-aveiga opened the pull request:
- Previous warnings when checking Composer configs are now errors.
- Skipping a check by adding configuration throws a warning instead.
- Adds one test workflow for one type of check.
  - This could be used as a template for other check tests, but I didn't add other tests to avoid putting more code on this PR.

beto-aveiga committed:
ISSUE-7: warnings instead of silence, errors instead of warnings... docs not updated yet

beto-aveiga committed:
ISSUE-7: fixing composer

beto-aveiga committed:
ISSUE-7: fixing composer (trying)

beto-aveiga committed:
ISSUE-7: avoid warnings if extra composer-checks is not set [skip ci]

beto-aveiga committed:
ISSUE-7: avoid warnings if other sub keys are not set

beto-aveiga committed:
ISSUE-7: fail if tehre is an error to communicate [skip ci]

beto-aveiga committed:
ISSUE-7: improve messages [skip ci]

beto-aveiga committed:
ISSUE-7: improve messages [skip ci]

beto-aveiga committed:
ISSUE-7: improve messages [skip ci]

beto-aveiga committed:
ISSUE-7: communicate the user how to avoid the error [skip ci]

beto-aveiga committed:
ISSUE-7: communicate the user how to avoid the error [skip ci]

beto-aveiga committed:
ISSUE-7: error by default [skip ci]

beto-aveiga committed:
ISSUE-7: change info messages to warnings for better user guidance [skip ci]

beto-aveiga committed:
ISSUE-7: more clarity when if the user decide to just show a warning for the checks [skip ci]

beto-aveiga committed:
ISSUE-7: updating composer-check configs to simpler values

beto-aveiga committed:
ISSUE-7: avoid telling the user how to to x,y,z... that is for the docs

beto-aveiga committed:
ISSUE-7: refator [skip ci]

beto-aveiga committed:
ISSUE-7: refactor how to define the message type sent tot he user [skip ci]

beto-aveiga committed:
ISSUE-7: updating documentation [skip ci]

beto-aveiga committed:
ISSUE-7: one small test, first try

beto-aveiga committed:
ISSUE-7: one small test, first try

beto-aveiga committed:
ISSUE-7: run real composer install

beto-aveiga committed:
ISSUE-7: user composer commands to generate composer.json

beto-aveiga committed:
ISSUE-7: one small test, first try

beto-aveiga committed:
ISSUE-7: update composer install commands to include --ignore-platform-reqs for better error handling

beto-aveiga committed:
ISSUE-7: easier debugging

beto-aveiga committed:
ISSUE-7: use the composer-check code in the runner

beto-aveiga committed:
ISSUE-7: improve error handling in composer install steps by temporarily disabling exit on error

beto-aveiga committed:
Added .gitignore file to exclude test files from version control

beto-aveiga committed:
Update .gitignore to exclude vendor directory in addition to test files

beto-aveiga committed:
ISSUE-7: fix composer configuration building

beto-aveiga committed:
ISSUE-7: fix test

beto-aveiga committed:
ISSUE-7: minor changes, adding lines at the end files

beto-aveiga committed:
ISSUE-7: adding empty line

deviantintegral committed:
Add missing composer/composer dev dependency

https://getcomposer.org/doc/articles/plugins.md#creating-a-plugin

deviantintegral reviewed (commented):
Thanks for this first pass! I pushed up a commit requiring `composer/composer` so local IDEs work correctly for class lookups. Otherwise, see my notes!

beto-aveiga commented on src/ComposerChecksPlugin.php:
```diff
@@ -76,26 +76,22 @@ public static function getSubscribedEvents()
     if ($patchesInExtraFile) {
       if (!file_exists($patchesInExtraFile)) {
-        $this->io->warning("The patches file `$patchesInExtraFile` can't be read.");
+        $this->io->$message_type("The patches file `$patchesInExtraFile` can't be read.");
```
I will update this, of course!

beto-aveiga commented on .github/workflows/test-disable-drupal-core-patches-level-check.yml:
```diff
@@ -0,0 +1,192 @@
+
+on:
+  push:
+    branches: [ main, develop ]
+  pull_request:
+    branches: [ main, develop ]
```
Yeah, we need this for all pull requests. I agree.
I was focused on the testing logic and forgot to check the workflow triggering rules.
Thanks!

beto-aveiga committed:
Update .github/workflows/test-disable-drupal-core-patches-level-check.yml

Co-authored-by: Andrew Berry <andrew.berry@lullabot.com>

beto-aveiga commented on .github/workflows/test-disable-drupal-core-patches-level-check.yml:
```diff
@@ -0,0 +1,192 @@
+      uses: shivammathur/setup-php@v2
+      with:
+        php-version: ${{ matrix.php-version }}
+        extensions: json, mbstring
```
I will remove them and re-run tests, and update based on the feedback.

beto-aveiga commented on .github/workflows/test-disable-drupal-core-patches-level-check.yml:
```diff
@@ -0,0 +1,192 @@
+        composer config repositories.lullabot/composer-checks '{"type": "path", "url": "../"}'
+        composer config allow-plugins.lullabot/composer-checks true
+        composer config allow-plugins.cweagans/composer-patches true
+        composer require drupal/core:^10.0 --no-update
```
Yeah, requiring Drupal may be too much; maybe it is unnecessary. I will double-check this.

beto-aveiga commented on .github/workflows/test-disable-drupal-core-patches-level-check.yml:
```diff
@@ -0,0 +1,192 @@
+        jq '.extra["composer-checks"][1] = "break-composer-installation-if-patches-fail"' composer.json > composer.json.tmp && mv composer.json.tmp composer.json
+        jq '.extra["composer-checks"][2] = "disable-exit-on-patch-failure-check"' composer.json > composer.json.tmp && mv composer.json.tmp composer.json
+
+    - name: Test 1 - Verify check fails without disable flag
+      run: |
```
All tests use different folders. So, one is not affected by the other tests.
Also, there is a step to build composer.json, and the checks are run in the next step.
Please let me know if I'm missing something.

beto-aveiga committed:
Update workflow to trigger on main branch only and every pull request

beto-aveiga committed:
Refactor logging method in ComposerChecksPlugin to use log() instead of dynamic method call for message types

beto-aveiga committed:
Update error handling in ComposerChecksPlugin to throw RuntimeException instead of exiting the process abruptly. This allow Composer to do whatever it needs to do after our command fails.

beto-aveiga committed:
Update PHP version matrix in workflow to include PHP 8.4

beto-aveiga committed:
These extesions may be unnecessary

beto-aveiga committed:
Remove unnecessary drupal/core dependency from workflow configuration

beto-aveiga committed:
Fix log calls by swapping parameters

beto-aveiga committed:
ISSUE-7: allow dev, prefer stable for composer

beto-aveiga committed:
Refactor logging calls in ComposerChecksPlugin to use log() instead dynamic method for warning and errors

beto-aveiga commented:
@deviantintegral: I updated a few things based on your comments and replied to all of them.

Please review this again when you can. Many thanks for the feedback!

deviantintegral committed:
Initial pass moving tests to PHPUnit with Junie.

The initial prompt rewrote the tests using normal phpunit style assertions instead of functional / CLI tests. With a second try being specific about that in the prompt it seemed to get the message.

Clear issues so far are that it used old unsupported versions of libraries.

The tests currently in `.github/workflows/test-disable-drupal-core-patches-level-check.yml` need to be rewritten to run with PHPUnit. The GitHub workflows should then run phpunit to test the library.

Modify the project to require phpunit as a dev dependency. Create new PHPUnit tests in a folder at `tests/src/Functional', with each test currently in the GitHub workflow created as a new test class. Create a basic `phpunit.xml.dist` file to support running these tests. Finally, update the github workflow to run the tests. Focus on porting the tests as directly as possible, without making unrelated improvements. The PHPUnit tests should use the Symfony process component to run the commands as currently shown in the workflow file. At the end, confirm that each line of shell scripting removed is accurately included in the new PHPUnit tests.

deviantintegral committed:
Run all tests instead of just the current one

deviantintegral committed:
Update to latest symfony/process

deviantintegral committed:
Update to PHPUnit 12

deviantintegral committed:
setUp() is called once per test, and remove insecure 0777 permissions

deviantintegral committed:
Rename test class as it is testing more than one case

deviantintegral committed:
Downgrade to support PHP 8.1

deviantintegral committed:
Typehint the testDir property

deviantintegral committed:
Remove unused return

deviantintegral committed:
Update schema to match version

deviantintegral committed:
Migrate configuration

deviantintegral committed:
Fix assuming being run in the project root directory

deviantintegral committed:
Simplify workflow file name

deviantintegral committed:
Simplify string syntax

deviantintegral committed:
Use paratest for faster execution

deviantintegral committed:
Update workflow names

deviantintegral committed:
Downgrade paratest to see if it works with old php

deviantintegral committed:
Use phpunit to see if it passes on old PHP versions

deviantintegral committed:
Disable fail-fast so we can see all PHP version tests

rabbitlair reviewed (approved):
Looks good to me!

Also, I agree with @deviantintegral, the ability to test locally with PHPUnit is worth to have.

beto-aveiga committed:
Merge pull request #13 from Lullabot/issue-7-phpunit

Issue 7 Switch to phpunit

rabbitlair reviewed (approved):
Looks good to me :100:
